sql/catalyst/src/main/scala/org/apache/spark/sql/types/DecimalType.scala

```diff
@@ -148,13 +152,28 @@ object DecimalType extends AbstractDataType {
   }
 
   private[sql] def fromBigDecimal(d: BigDecimal): DecimalType = {
-    DecimalType(Math.max(d.precision, d.scale), d.scale)
+    fromJavaBigDecimal(d.underlying())
+  }
+
+  private[sql] def fromJavaBigDecimal(d: JavaBigDecimal): DecimalType = {
+    val (precision, scale) = if (d.scale < 0 && SQLConf.get.ansiEnabled) {
+      (d.precision - d.scale, 0)
+    } else {
+      (d.precision, d.scale)
+    }
+    DecimalType(Math.max(precision, scale), scale)
```

Comment:
do we still need `Math.max(precision, scale)`? seems it's always `precision` now.